ydb/core/tx/datashard/import_s3.cpp

```diff
@@ -656,21 +652,35 @@ class TS3Downloader: public TActorBootstrapped<TS3Downloader> {
         return true;
     }
 
-    void RestartOrFinish(const TString& error) {
-        if (Attempt++ < Retries) {
-            Delay = Min(Delay * Attempt, MaxDelay);
-            const TDuration random = TDuration::FromValue(TAppData::RandomProvider->GenRand64() % Delay.MicroSeconds());
+    static bool ShouldRetry(const Aws::S3::S3Error& error) {
+        return error.ShouldRetry();
+    }
 
-            Schedule(Delay + random, new TEvents::TEvWakeup(static_cast<ui64>(EWakeupTag::Restart)));
-        } else {
-            Finish(false, error);
-        }
+    static bool ShouldRetry(const TString&) {
+        return true;
     }
 
-    void Handle(TEvents::TEvWakeup::TPtr& ev) {
-        switch (static_cast<EWakeupTag>(ev->Get()->Tag)) {
-        case EWakeupTag::Restart:
-            return Restart();
+    template <typename T>
+    bool CanRetry(const T& error) const {
+        return Attempt < Retries && ShouldRetry(error);
+    }
+
+    void Retry() {
+        Delay = Min(Delay * ++Attempt, MaxDelay);
+        const TDuration random = TDuration::FromValue(TAppData::RandomProvider->GenRand64() % Delay.MicroSeconds());
+        Schedule(Delay + random, new TEvents::TEvWakeup());
+    }
+
+    template <typename T>
+    void RetryOrFinish(const T& error) {
+        if (CanRetry(error)) {
+            Retry();
+        } else {
+            if constexpr (std::is_same_v<T, Aws::S3::S3Error>) {
+                Finish(false, TStringBuilder() << "S3 error: " << error.GetMessage().c_str());
+            } else {
+                Finish(false, error);
+            }
```

Comment:
Why did you use nested if here? `constexpr if` can be combined with regular `if`.